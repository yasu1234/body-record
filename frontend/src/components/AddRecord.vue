<script setup>
import { ref, onMounted } from 'vue';
import axios from 'axios';
import Cookies from 'js-cookie';

import DatePicker from './DatePicker.vue'
import DropFile from './DropFile.vue'

const content = ref("");
const recordDate = ref("");
const files = ref([]);

function dateChange(event) {
    recordDate.value = event
}

function onFileChange(event) {
    files.value = [...event];
}

const registerRecord = async () => {
  try {
        const formData = new FormData();
        formData.append('memo', content);
        formData.append('date', recordDate);

        for (const file of files.value) {
            formData.append('images', file);
        }

        const res = await axios.post(import.meta.env.VITE_APP_API_BASE + '/api/v1/records', formData, {
            headers: {
                'Content-Type': 'multipart/form-data',
                'access-token' : Cookies.get('accessToken'),
                'client':Cookies.get('client'),
                'uid': Cookies.get('uid')
            }
        })
        console.log({ res })
    } catch (error) {
        console.log({ error })
    }
}
</script>

<template>
    <div class="time-list">
        <div class="item">
            <p class="inputTitle">記録日</p>
            <DatePicker isStart=false @startDateChange="dateChange"/>
        </div>
    </div>
    <div class="editor">
        <input type="text" id="Name" name="Name" placeholder="内容" v-model="content">
    </div>
    <div class="relationImages">
        <p class="inputTitle">関連画像</p>
        <DropFile @change="onFileChange"/>
    </div>
    <div class="relationImages">
        <button class="registerButton" @click="registerRecord">登録する</button>
    </div>
</template>

<style>
.editor{
   padding: 30px;
 }

 input[type=text] {
  width: 100%;
  padding: 12px 12px;
  margin: 8px 0;
  box-sizing: border-box;
  border: 1px solid #ccc;
  border-radius: 4px;
}
.time-list {
  display: flex;
  justify-content: space-between;
  padding-left: 30px;
  padding-right: 30px;
}
.time-list .item {
  width: 50%;
  margin: 0;
  padding: 10px;
  box-sizing: border-box;
}
.time-list .item .inputTitle {
  margin: 5px 0 0;
  padding: 0;
  font-size: 16px;
}
.relationImages {
    padding: 20px;
}
.registerButton{
    background: #ffa500;
    color: white;
    font-size:16px;
    font-weight:bold;
}
</style>