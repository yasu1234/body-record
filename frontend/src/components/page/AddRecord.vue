<script setup>
import { ref, onMounted } from "vue";
import axios from "axios";
import Cookies from "js-cookie";
import Textarea from "primevue/textarea";
import FloatLabel from "primevue/floatlabel";
import InputText from "primevue/inputtext";

import DatePicker from "../atom/DatePicker.vue";
import DropFile from "../atom/DropFile.vue";
import Header from "../layout/Header.vue";
import ErrorMessage from "../atom/ErrorMessage.vue";

const memo = ref("");
const recordDate = ref("");
const weight = ref(null);
const fatPercentage = ref(null);
const files = ref([]);
const isAddAsHidden = ref(false);
const errorMessage = ref("");

function dateChange(event) {
  recordDate.value = event;
}

function onFileChange(event) {
  files.value = [...event];
}

const registerRecord = async () => {
  try {
    const formData = new FormData();
    formData.append("record[memo]", memo.value);
    formData.append("record[date]", recordDate.value);
    formData.append("record[weight]", weight.value);
    formData.append("record[fat_percentage]", fatPercentage.value);

    if (isAddAsHidden.value === false) {
      formData.append("record[open_status]", 1);
    }

    for (const file of files.value) {
      formData.append("images", file);
    }

    const res = await axios.post(
      import.meta.env.VITE_APP_API_BASE + "/api/v1/records",
      formData,
      {
        headers: {
          "Content-Type": "multipart/form-data",
          "access-token": Cookies.get("accessToken"),
          client: Cookies.get("client"),
          uid: Cookies.get("uid"),
        },
      }
    );
    console.log({ res });
  } catch (error) {
    errorMessage.value = "";
    let errorMessages = "記録の追加に失敗しました\n";
    if (error.response.status === 422) {
      if (Array.isArray(error.response.data.errors)) {
        errorMessages += error.response.data.errors.join("\n");
      }
    }
    errorMessage.value = errorMessages;
  }
};
</script>

<template>
  <Header />
  <ErrorMessage :errorMessage="errorMessage" />
  <div class="add-record-container">
    <div class="record-add-space">
      <p>記録日</p>
      <DatePicker
        isStart="false"
        :date="recordDate"
        @update:date="dateChange"
        class="input-width"
      />
    </div>
    <div class="record-add-space input-group">
      <FloatLabel>
        <InputText v-model="weight" class="input-width" />
        <label>体重</label>
      </FloatLabel>
      <label for="goal-weight" class="unit-label">kg</label>
    </div>
    <div class="record-add-space input-group">
      <FloatLabel>
        <InputText v-model="fatPercentage" class="input-width" />
        <label>体脂肪率</label>
      </FloatLabel>
      <label for="goal-fat-percentage" class="unit-label">%</label>
    </div>
    <div class="record-add-space">
      <FloatLabel>
        <Textarea v-model="memo" rows="10" />
        <label>メモ</label>
      </FloatLabel>
    </div>
    <div class="record-add-space">
      <input type="checkbox" id="statusSelect" v-model="isAddAsHidden" />
      <label for="statusSelectName">非公開記録にする場合にはチェック</label>
    </div>
    <div class="record-add-space">
      <p>関連画像</p>
      <DropFile @change="onFileChange" />
    </div>
  </div>
  <div class="record-add-space">
    <button class="add-record-button" @click="registerRecord">記録を登録する</button>
  </div>
</template>

<style scoped>
.add-record-container {
  margin-top: 20px;
  margin-left: 30px;
}
.input-group {
  display: flex;
  align-items: center;
}
.input-width {
  width: 200px;
}
.record-add-space {
  margin-top: 30px;
}
.add-record-button {
  font-size: 16px;
  font-weight: bold;
  padding: 10px 50px;
}
</style>
