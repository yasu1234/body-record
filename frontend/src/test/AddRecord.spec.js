import { shallowMount } from '@vue/test-utils'
import AddRecord from '@/components/page/AddRecord.vue'

jest.mock('axios')
jest.mock('vue-router', () => {
    const router = {
      push: jest.fn(),
      pathname: '/',
      route: '',
      asPath: '',
      query: {},
      events: {
        on: jest.fn(),
        off: jest.fn(),
        emit: jest.fn(),
      },
    };
    return {
      useRouter: jest.fn().mockReturnValue(router)
    };
  });

  jest.mock('primevue/usetoast', () => ({
    useToast: jest.fn(() => ({
      add: jest.fn(),
      removeAll: jest.fn()
    }))
  }))

describe('AddRecord', () => {
  let wrapper

  beforeEach(() => {
    wrapper = shallowMount(AddRecord, {
        global: {
          stubs: ["router-link", "router-view"]
        }
    })
  })

  afterEach(() => {
    jest.clearAllMocks()
  })

  it('SnapShotテスト初期表示', () => {
    expect(wrapper.element).toMatchSnapshot();
  })
})
