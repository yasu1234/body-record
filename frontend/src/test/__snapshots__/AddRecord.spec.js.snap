// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddRecord SnapShotテスト初期表示 1`] = `
<div
  data-v-app=""
>
  
  <header-stub />
  <tab-menu-stub />
  <toast-stub
    position="top-center"
  />
  <form
    class="add-record-container"
  >
    <div
      class="record-add-space"
    >
      <p>
        記録日
      </p>
      <date-picker-stub
        class="input-width"
        date=""
        isstart="false"
      />
    </div>
    <div
      class="input-group mt-7"
    >
      <float-label-stub />
      <label
        class="unit-label"
        for="goal-weight"
      >
        kg
      </label>
    </div>
    <div
      class="input-group mt-7"
    >
      <float-label-stub />
      <label
        class="unit-label"
        for="goal-fat-percentage"
      >
        %
      </label>
    </div>
    <div
      class="mt-7"
    >
      <float-label-stub />
    </div>
    <div
      class="mt-7"
    >
      <input
        class="w-4 h-4 text-blue-600 bg-gray-100 border-gray-300 rounded focus:ring-blue-500"
        id="statusSelect"
        type="checkbox"
      />
      <label
        for="statusSelectName"
      >
        非公開記録にする場合にはチェック
      </label>
    </div>
    <div
      class="mt-7"
    >
      <p>
        関連画像
      </p>
      <drop-file-stub />
    </div>
    <div
      class="record-button-space mt-7"
    >
      <button
        class="add-record-button"
      >
        記録を登録する
      </button>
    </div>
  </form>
  
</div>
`;
